x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java

```diff
@@ -1249,60 +1255,137 @@ public String getFeatureDescription() {
+
+        ActionListener<ResetFeatureStateResponse.ResetFeatureStateStatus> unsetResetModeListener = ActionListener.wrap(
+            success -> client.execute(SetResetModeAction.INSTANCE, SetResetModeAction.Request.disabled(), ActionListener.wrap(
+                resetSuccess -> finalListener.onResponse(success),
+                resetFailure -> {
+                    logger.warn("failed to disable reset mode after state clean up success", resetFailure);
+                    finalListener.onResponse(success);
+                })
+            ),
+            failure -> client.execute(SetResetModeAction.INSTANCE, SetResetModeAction.Request.disabled(), ActionListener.wrap(
+                resetSuccess -> finalListener.onFailure(failure),
+                resetFailure -> {
+                    logger.warn("failed to disable reset mode after state clean up failure", resetFailure);
+                    finalListener.onFailure(failure);
+                })
+            )
+        );
 
         Map<String, Boolean> results = new ConcurrentHashMap<>();
 
+        ActionListener<ListTasksResponse> afterWaitingForTasks = ActionListener.wrap(
+            listTasksResponse -> {
+                listTasksResponse.rethrowFailures("Waiting for indexing requests for .ml-* indices");
+                if (results.values().stream().allMatch(b -> b)) {
+                    // Call into the original listener to clean up the indices
+                    SystemIndexPlugin.super.cleanUpFeature(clusterService, client, unsetResetModeListener);
+                } else {
+                    final List<String> failedComponents = results.entrySet().stream()
+                        .filter(result -> result.getValue() == false)
+                        .map(Map.Entry::getKey)
+                        .collect(Collectors.toList());
+                    unsetResetModeListener.onFailure(new RuntimeException("Some components failed to reset: " + failedComponents));
+                }
+            },
+            unsetResetModeListener::onFailure
+        );
+
         ActionListener<StopDataFrameAnalyticsAction.Response> afterDataframesStopped = ActionListener.wrap(dataFrameStopResponse -> {
             // Handle the response
             results.put("data_frame/analytics", dataFrameStopResponse.isStopped());
-
             if (results.values().stream().allMatch(b -> b)) {
-                // Call into the original listener to clean up the indices
-                SystemIndexPlugin.super.cleanUpFeature(clusterService, client, listener);
+                client.admin()
+                    .cluster()
+                    .prepareListTasks()
+                    .setActions("xpack/ml/*")
+                    .setWaitForCompletion(true)
+                    .execute(ActionListener.wrap(
+                        listMlTasks -> {
+                            listMlTasks.rethrowFailures("Waiting for machine learning tasks");
```

Comment:
Is the idea to leave the plugin in reset mode after a timeout or something so it can be tried again?

Reply:
No, the idea is reset mode to be unset. 

See https://github.com/elastic/elasticsearch/blob/318ae89e14b9b1b925d33e139177a8e51297d12c/server/src/main/java/org/elasticsearch/action/ActionListener.java#L128-L150